### PR TITLE
Use Index rather than Long for IntList parsing

### PR DESCRIPTION
**Pull request by gchanan**
so floating-point types convertible to ints fail the parsing.

Basically, our unpackLong code works with floating-point types that are convertible to ints, but this isn't often what you want (because of truncation).
What you actually want is to convert to an index, which will usually flag such issues.

I made this the minimal change I could because:
1) I didn't want to change unpackLong because the existing code call checkLong before unpackLong, so this should be a non-issue most of the time.  And fixing unpackLong requires calling checkLong again, which will slow everything down.
2) An exception above is with IntList, which only checks that 1) it is a tuple or 2) it is a varargs tuple (i.e. torch.ones(1, 2, 3)).

### CODE REVIEW

**Comment by gchanan** [comment]
Related issue: https://github.com/pytorch/pytorch/issues/6663

**Comment by gchanan** [comment]
Note this is roughly what Numpy does with integer tuples: https://github.com/numpy/numpy/blob/8323be1bc44c2811fc36f5b99c1a30ebcee8edbd/numpy/core/src/multiarray/conversion_utils.c#L792-L882

The difference is they use PyLong_CheckExact / PyInt_CheckExact instead of PyLong / PyInt.  We could probably do that to because I'm guessing the PyNumber_Index call will work for non-exact PyLongs / PyInts.

**Review by colesbury — APPROVED** [review]

File: torch/csrc/utils/python_numbers.h [hunk 1]
```diff
@@ -57,6 +57,29 @@ inline int64_t THPUtils_unpackLong(PyObject* obj) {
   return (int64_t)value;
 }
 
+inline bool THPUtils_checkIndex(PyObject *obj) {
+  if (THPUtils_checkLong(obj)) {
+    return true;
+  }
```
Comment (colesbury) [inline torch/csrc/utils/python_numbers.h]: I think this should have something like:

```
if (PyBool_Check(Obj)) return false
```

Comment (gchanan) [reply torch/csrc/utils/python_numbers.h]: sounds right.

**Comment by gchanan** [comment]
looks like my merge screwed up the submodules, going to try again.

**Pull request merged**